web/src/message_list.js

```diff
@@ -76,10 +80,27 @@ export class MessageList {
         // This is intentionally not live-updated when web_home_view
         // changes, since it's easier to reason about if this
         // optimization is active or not for an entire session.
-        this.preserve_rendered_state =
-            user_settings.web_home_view === "all_messages" && !this.narrowed;
+        if (user_settings.web_home_view !== "all_messages" || this.narrowed) {
+            return false;
+        }
 
-        return this;
+        // Since our natural tendency is to narrow user to `first_unread` message,
+        // we don't preserve the rendered state of this list if first unread message
+        // is not available to us. Otherwise, this leads to confusion when we are
+        // restoring the rendered list but our first unread message is not available
+        // and fetching it from the server could lead to non-contiguous messages history.
+        //
+        // NOTE: Non-contiguous message history can still happen in the opposite situation
+        // where user is narrowing to a message id which is not present in the rendered list.
+        // In this case, we create a new list and if the new fetched history contains first
+        // unread message, we preserver this list and discard others with the same filter.
+        const first_unread_message = this.get(this.data.first_unread_message_id());
+        if (!first_unread_message?.unread) {
+            // If we have found the newest message, we can preserve the rendered state.
+            return this.data.fetch_status.has_found_newest();
+        }
+
+        return true;
```

Comment:
I don't think this logic is correct; we need to be looking up the very first unread message that the user has matching the combined feed's filters, which may not be in `this.data` at all -- so I think the `this.data.first_unread_message_id()` check needs to be instead querying `unread.ts` to ask (1) whether `old_unreads_missing` is true; and if not, what is the oldest non-muted unread message ID -- and then we can just check if that message ID is already present in `this.data`, or equivalently, whether it's within the range of messages fetched for `this.data`.

In theory, we should already have that number in the `narrow.js` code that calls this, so you might not need any code here to find the ID so much as to just pass the resolved `id_info.final_select_id` from the calling code.

Might require some refactoring to basically move the `get_first_unread_info` call out of `maybe_add_local_messages`, or something, but I think such refactoring is a significant chunk of useful work towards generalizing this logic to support https://github.com/zulip/zulip/issues/15131 -- basically we'd extend the existing loop over available cached `MessageListData` objects that decides whether to use a cached something to find the one that has the best superset of the target narrow's data available, where best is something like "Conditioned on it having the target/anchor message in its contiguous region, has the most history".

Reply:
I see, that is indeed bugged.